Products/ZPerFactMods/requestWrapper.py

```diff
@@ -0,0 +1,89 @@
+#
+# Hook into event handlers to call request_init/request_end
+#
+
+import zope.component
+import ZPublisher.pubevents
+import six
+import logging
+
+logger = logging.getLogger('Products.ZPerFactMods.requestWrapper')
+
+
+def get_context(event):
+    """
+    Look up the context of the published object. Returns None for some
+    irregular requests, for example for builtin ZMI assets that have no title
+    or when publishing a File object.
+    Also returns None if the published object's title contains NO_REQWRAP.
+    If the context does not contain all necessary acquisition components as
+    defined by the nav_acquire property of the site object returned by
+    get_site(), returns the site object instead.
+    """
+    try:
+        title = event.request.PUBLISHED.title
+        if callable(title):
+            title = title()
+        if isinstance(title, six.binary_type) and b'NO_REQWRAP' in title:
+            return None
+        if isinstance(title, six.text_type) and u'NO_REQWRAP' in title:
+            return None
+
+        context = event.request.PARENTS[0]
+        parents = {obj.id for obj in event.request.PARENTS}
+        root = context.get_site()
+        nav_acquire = set(root.nav_acquire)
+
+        if len(nav_acquire.difference(parents)):
+            # Parts from nav_acquire are missing
+            return root
+        return context
+
+    except Exception:
+        # Hook not implemented, published object has no title, ...
+        return None
+
+
+@zope.component.adapter(ZPublisher.pubevents.PubAfterTraversal)
+def request_init(event):
+    """
+    Call request_init before calling the published object. If it is not found,
+    call layout_init as compatibility fallback, but adjust the layer counter
```

Comment:
Adjusting the layer counter is really a smart hack.

Reply:
It comes at the cost of every request that already executed `layout_init` and `layout_end` explicitly now executing it twice, without short-circuiting (or rather with less short-circuiting. For example, `appsession_init` does its own recognition of already having been executed). But I think this is the best solution in the long run.